### ISSUE TITLE
Understand that a condition applies to the closure

### ISSUE THREAD

**Issue by williamdes** [nunomaduro/larastan#1081]
- Larastan Version: 8.77
- `--level` used: max

### Description

```
 Parameter #1 $user of static method                                    
         Util::doFoo() expects UserTypeSpecific, UserTypeSpecific|OtherUserType given. 
```

### Laravel code where the issue was found

```php
public function doFoo(UserTypeSpecific $type)
{
// code
}
/** @var UserTypeSpecific|OtherUserType */
$user = ...;
            $this->mergeWhen($user instanceof UserTypeSpecific, function() use ($user) {
                // should understand that at this point $user is a UserTypeSpecific
                $fooBar = Util::doFoo($user);
                return [
                    'baz' => $fooBar,
                ];
            }),
```

**Comment by szepeviktor** [comment]
> Larastan Version: 8.77

This is **Larastan** :)

**Comment by williamdes** [comment]
> > Larastan Version: 8.77
> 
> This is **Larastan** :)

Uuups, fixed
I read Laravel
maybe it would be interesting to be added to the template

**Comment by canvural** [comment]
Hi,

I don't think this is possible with PHPStan. You can try first opening an issue there and if it's possible, we can implement here.

**Comment by williamdes** [comment]
> Hi,
> 
> I don't think this is possible with PHPStan. You can try first opening an issue there and if it's possible, we can implement here.

Thanks
@ondrejmirtes is it worth I open an issue at phpstan ?

**Comment by ondrejmirtes** [comment]
I'd rather have it rewritten to be more static analyser-friendly code. I can think of a dozen ways of rewriting that :) https://phpstan.org/writing-php-code/narrowing-types

**Comment by williamdes** [comment]
> I'd rather have it rewritten to be more static analyser-friendly code. I can think of a dozen ways of rewriting that :) https://phpstan.org/writing-php-code/narrowing-types

Yeah but `$this->mergeWhen($user instanceof UserTypeSpecific, function() use ($user) {` is a laravel syntax

**Comment by williamdes** [comment]
https://laravel.com/docs/9.x/eloquent-resources#merging-conditional-attributes
